lib/spack/spack/build_systems/cmake.py

```diff
@@ -174,6 +174,70 @@ def _std_args(pkg):
         args.append('-DCMAKE_PREFIX_PATH:STRING={0}'.format(';'.join(deps)))
         return args
 
+    def define_from_variant(self, variant, cmake_var=None):
```

Comment:
It is somewhat irritating that the variant name comes before the variable name, and it's a bit implicit what happens when you just put `self.define_from_variant("foo")`.  What if we generalize the function a big and make it look like this:

```python
def define(self, cmake_var, value=None, variant=None):
    ...
```

Then you can generate a regular define with some string value like this:

```python
args.appen(self.define("FOO", "bar"))
```

but you can do the variant magic you implemented like this:

```python
args.appen(self.define("FOO", variant="foo"))
```

In both cases you can see what the variable name is by reading the package, which I think is a bit clearer.  If you really want the implicit define name, I suppose you could make the initial `cmake_var` parameter optional and allow `self.define(variant="foo")`.

Thoughts?

Reply:
Sounds great. That's actually closer to what I had originally -- but then I had doubts about putting the cmake variable first. And I really like allowing the general `value` this way too.

Reply:
Actually I'm not a fan of the "exactly one of these keyword arguments" idiom -- it's not intuitive from the function signature. Perhaps adding `define` with required key and value; and a `` utility function that (with error checking) calls
```python
def define_from_variant(self, cmake_var, variant):
    # check `variant` is valid, then...
    return self.define(key, self.spec.variants[variant].value)
```
?

Reply:
Or maybe just letting the user define a local helper function `def get_variant(variant): return self.spec.variants[variant].value` would be sufficient, so it could be used as
```python
args = [self.define('FOO', get_variant('foo')), ...]
```
and adding that mini-recipe to the documentation I'll add to `build_systems/cmakepackage.rst`.

Reply:
@tgamblin and I met in "real life" at ECP and it sounded like he approves of merging the current PR, so I'm marking this as resolved.